File: src/listunitareas.cpp
/*
 * The QED Programming Language
 * Copyright (C) 2022-2023  Hocus Codus Software inc.
 *
 * All rights reserved.
 */
#include <string.h>
#include "listunitareas.hpp"

LocationUnit::~LocationUnit() {
}
/*
  Object addValue(List list, Object value) {
    if (getNumLevels() != 0) {
      if (list == null) list = List.empty(growable: true);

      Object oldValue = get(0) < list.length ? list[get(0)] : null;
      Object newValue = trimRange(0, 1).addValue(oldValue, value);

      if (oldValue == null && get(0) != list.length) oldValue = oldValue;
      if (oldValue != newValue) list.add(newValue);

      return list;
    } else
      return value;
  }
*/
LocationUnit *LocationUnit::addValue(LocationUnit **start, const Path &path, LocationUnit *value) {
  if (path.size() != 0) {
    LocationTree *tree = *start ? (LocationTree *) *start : new LocationTree(1);

    *start = tree;

    LocationTree *oldValue = path[0] < tree->size() ? (LocationTree *) tree->operator[](path[0]) : NULL;
    LocationUnit *newValue = addValue((LocationUnit **) &oldValue, path.trim(0, 1), value);

    if (oldValue != newValue)
      oldValue->push_back(newValue);

    return oldValue;
  }
  else
    return value;
}

LocationUnit *LocationUnit::getValue(const Path &path) {
  return this;
}

LocationTree::LocationTree(int count) {
}

LocationUnit *LocationTree::getValue(const Path &path) {
  return path.size() ? (LocationTree *) operator[](path[0])->getValue(path.trim(0, 1)) : this;
}

int ListUnitAreas::getSize(int dir) {
  return 0;
}

UnitArea::UnitArea(const Point &unitArea) {
  this->unitArea = unitArea;
}

int UnitArea::getSize(int dir) {
  return unitArea[dir];
}

IntTreeUnit::IntTreeUnit(int value) {
  this->value = value;
}

IntTreeUnit::IntTreeUnit(int value, IntTreeUnit *children) : std::vector<IntTreeUnit *>(*children) {
  this->value = value;
}

void IntTreeUnit::add(IntTreeUnit *unit) {
  push_back(unit);
}

int IntTreeUnit::getNumChildren() {
  return size();
}

LayoutUnitArea::LayoutUnitArea(const DirIntTreeUnits &trees, void *unitAreas) {
  this->trees = trees;
  this->unitAreas = unitAreas;
}

int LayoutUnitArea::getSize(int dir) {
  return trees[dir]->value;
}
/*
ValueStackElement::ValueStackElement(long parseFlags, bool b) {
  this->parseFlags = parseFlags;
  this->b = b;
}
*/

File: ui/ListUnitAreas.dart
/*
 *    Copyright (C) 2020 Hocus Codus Software inc. All rights reserved.
 *
 *    Author: Martin Savage (hocuscodus@gmail.com)
 */

import 'dart:collection';

class ListUnitAreas {
	int getSize(int dir) {
		return 0;
	}
}

class UnitArea extends ListUnitAreas {
	List<int> unitArea;

	UnitArea(List<int> unitArea) {
		this.unitArea = unitArea;
	}

	int getSize(int dir) {
		return unitArea[dir];
	}
}

class LayoutUnitArea extends ListUnitAreas {
	List<IntTreeUnit> trees;
	List unitAreas;

	LayoutUnitArea(List<IntTreeUnit> trees, List unitAreas) {
		this.trees = trees;
		this.unitAreas = unitAreas;
	}

	int getSize(int dir) {
		return trees[dir].value;
	}
}

class Stack<T> {
	final _stack = Queue<T>();

	void push(T element) {
		_stack.addLast(element);
	}

	T pop() {
		final T lastElement = _stack.last;
		_stack.removeLast();
		return lastElement;
	}

	T getValue() {
		return _stack.last;
	}

	void clear() {
		_stack.clear();
	}

	bool get isEmpty => _stack.isEmpty;
}

class ValueStack {
	HashMap<String, Stack<Object>> map = new HashMap<String, Stack<Object>>();

	void push(String key, Object value) {
		if (!map.containsKey(key))
			map[key] = new Stack<Object>();

		map[key].push(value);
	}

	Object pop(String key) {
		var rc = map[key].pop();

		if (map[key].isEmpty)
			map.remove(key);

		return rc;
	}

	Object get(String key) {
		return map[key].getValue();
	}
}

class IntTreeUnit {
	List<IntTreeUnit> set;
	int value;

	IntTreeUnit(int value) {
		set = List.filled(0, null, growable: true);
		this.value = value;
	}

	IntTreeUnit.fromList(int value, List<IntTreeUnit> children) {
		set = children;
		this.value = value;
	}

	void add(IntTreeUnit unit) {
		set.add(unit);
	}

	int getNumChildren() {
		return set != null ? set.length : 0;
	}
}
